### PR TITLE
feat(graphql): add support for subscriptions

### PR DESCRIPTION
**Pull request by S-furi**
This PR contains the necessary code for running the server and run upon it GraphQL operations about the underlying Alchemist simulation. New mutation (queries with side effects) has been added for managing the simulation status (i.e. play, pause, terminate). Two subscription as a proof of concept has been also added, mainly for receiving real time updates on the environment or a specific node given its Id.

### CODE REVIEW

File: alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/monitor/EnvironmentSubscriptionMonitor.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        this.stepDone(environment, null, Time.ZERO, 0L)
+    }
+
+    /**
+     * Returns a [Flow] that emits a new [EnvironmentSurrogate] each time the
+     * [OutputMonitor.stepDone] function is called.
+     */
+    fun subscribe(): Flow<EnvironmentSurrogate<T, P>> {
+        return internalFlow
+    }
```
Comment (DanySK) [inline alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/monitor/EnvironmentSubscriptionMonitor.kt:54]:
```suggestion
    fun subscribe(): Flow<EnvironmentSurrogate<T, P>> = internalFlow
```

This is not a subscription, though. A subscription takes a target and return `Unit`, this is an eventFlow

**Review by DanySK — CHANGES_REQUESTED** [review]
Reviewed 5/12

File: alchemist-graphql/build.gradle.kts [hunk 2]
```diff
@@ -79,6 +79,9 @@ kotlin {
                 implementation(rootProject)
                 implementation(libs.bundles.graphql.server)
                 implementation(libs.bundles.ktor.server)
+                implementation(incarnation("sapere"))
```
[the remark this answers is not in the document]
Comment (S-furi) [reply alchemist-graphql/build.gradle.kts]: Yes, I have added it for running the server with a test simulation. I will remove it!

File: alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/monitor/EnvironmentSubscriptionMonitor.kt (hunk 1 above)
Comment (S-furi) [reply alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/monitor/EnvironmentSubscriptionMonitor.kt]: You are right! Fixed it!

File: alchemist-graphql/build.gradle.kts [hunk 3]
```diff
@@ -79,6 +79,9 @@ kotlin {
                 implementation(rootProject)
                 implementation(libs.bundles.graphql.server)
                 implementation(libs.bundles.ktor.server)
+//                implementation(incarnation("sapere"))
```
Comment (DanySK) [inline alchemist-graphql/build.gradle.kts:82]:
```suggestion
```

Comment (DanySK) [reply alchemist-graphql/build.gradle.kts:82]: you may want to `testRuntime` it

File: alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/monitor/EnvironmentSubscriptionMonitor.kt [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ * @param P the position
+ */
+class EnvironmentSubscriptionMonitor<T, P : Position<out P>> : OutputMonitor<T, P> {
+    private val internalFlow = MutableSharedFlow<EnvironmentSurrogate<T, P>>(
+        replay = 1,
+        onBufferOverflow = BufferOverflow.DROP_OLDEST,
+    )
+
+    override fun stepDone(environment: Environment<T, P>, reaction: Actionable<T>?, time: Time, step: Long) {
+        sendEnvironment(environment)
```
Comment (DanySK) [inline alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/monitor/EnvironmentSubscriptionMonitor.kt:37]:
```suggestion
        internalFlow.tryEmit(environment.toGraphQLEnvironmentSurrogate())
```

File: alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/monitor/EnvironmentSubscriptionMonitor.kt [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+    override fun initialized(environment: Environment<T, P>) {
+        this.stepDone(environment, null, Time.ZERO, 0L)
+    }
+
+    /**
+     * Returns a [Flow] that emits a new [EnvironmentSurrogate] each time the
+     * [OutputMonitor.stepDone] function is called.
+     */
+    fun eventFlow(): Flow<EnvironmentSurrogate<T, P>> = internalFlow
```
Comment (DanySK) [inline alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/monitor/EnvironmentSubscriptionMonitor.kt:52]: What is the point of a `fun`?
```suggestion
    val eventFlow: Flow<EnvironmentSurrogate<T, P>> get() = internalFlow
```

File: alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/monitor/EnvironmentSubscriptionMonitor.kt [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * Sends the given [Environment].
+     * @param environment the environment to send.
+     */
+    private fun sendEnvironment(environment: Environment<T, P>) {
+        internalFlow.tryEmit(environment.toGraphQLEnvironmentSurrogate())
+    }
```
Comment (DanySK) [inline alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/monitor/EnvironmentSubscriptionMonitor.kt:60]:
```suggestion
```

File: alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/schema/operations/mutations/SimulationManagerMutations.kt [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+import it.unibo.alchemist.model.Environment
+import it.unibo.alchemist.model.Position
+
+/**
+ * A [SimulationManagerMutations] provides GraphQL [Mutation] for manipulating the current simulation.
+ */
+class SimulationManagerMutations<T, P : Position<out P>>(
+    private val environment: Environment<T, P>,
+) : Mutation {
+    /**
```
Comment (DanySK) [inline alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/schema/operations/mutations/SimulationManagerMutations.kt:24]:
```suggestion

    /**
```

File: alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/schema/operations/mutations/SimulationManagerMutations.kt [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import com.expediagroup.graphql.generator.annotations.GraphQLDescription
+import com.expediagroup.graphql.server.operations.Mutation
+import it.unibo.alchemist.core.Status
+import it.unibo.alchemist.model.Environment
+import it.unibo.alchemist.model.Position
+
+/**
+ * A [SimulationManagerMutations] provides GraphQL [Mutation] for manipulating the current simulation.
+ */
+class SimulationManagerMutations<T, P : Position<out P>>(
```
Comment (DanySK) [inline alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/schema/operations/mutations/SimulationManagerMutations.kt:21]:
```suggestion
class SimulationControl<T, P : Position<out P>>(
```

File: alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/schema/operations/mutations/SimulationManagerMutations.kt [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    private val environment: Environment<T, P>,
+) : Mutation {
+    /**
+     * Play the simulation.
+     */
+    @GraphQLDescription("Play the simulation")
+    fun playSimulation(): String =
```
Comment (DanySK) [inline alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/schema/operations/mutations/SimulationManagerMutations.kt:28]:
```suggestion
    fun play(): String =
```

File: alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/schema/operations/mutations/SimulationManagerMutations.kt [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+            .fold(
+                onSuccess = { Status.RUNNING.toString() },
+                onFailure = { it.message.toString() },
+            )
+
+    /**
+     * Pause the simulation.
+     */
+    @GraphQLDescription("Pause the simulation")
+    fun pauseSimulation(): String =
```
Comment (DanySK) [inline alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/schema/operations/mutations/SimulationManagerMutations.kt:39]:
```suggestion
    fun pause(): String =
```

File: alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/schema/operations/mutations/SimulationManagerMutations.kt [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+            .fold(
+                onSuccess = { Status.PAUSED.toString() },
+                onFailure = { it.message.toString() },
+            )
+
+    /**
+     * Terminate the simulation.
+     */
+    @GraphQLDescription("Terminate the simulation")
+    fun terminateSimulation(): String =
```
Comment (DanySK) [inline alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/schema/operations/mutations/SimulationManagerMutations.kt:50]:
```suggestion
    fun terminate(): String =
```

File: alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/schema/operations/mutations/SimulationManagerMutations.kt [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        runCatching { this.environment.simulation.apply { terminate() } }
+            .fold(
+                onSuccess = { Status.TERMINATED.toString() },
+                onFailure = { it.message.toString() },
+            )
+}
```
Comment (DanySK) [inline alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/graphql/schema/operations/mutations/SimulationManagerMutations.kt:56]: The implementation of all methods is the same, and we don't do copy/paste here :)
refactor please

File: alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/launchers/GraphQLServerLauncher.kt [hunk 13]
```diff
@@ -31,16 +32,17 @@ class GraphQLServerLauncher @JvmOverloads constructor(
...
         val simulation: Simulation<Any, Nothing> = prepareSimulation(loader, emptyMap<String, Any>())
+        simulation.addOutputMonitor(EnvironmentSubscriptionMonitor())
         startServer(simulation)
     }
 
-    private fun <T, P : Position<P>> startServer(
+    private fun <T, P : Position<out P>> startServer(
         simulation: Simulation<T, P>,
-        serverDispatcher: CoroutineDispatcher = Dispatchers.IO,
+        serverDispatcher: CoroutineDispatcher = Dispatchers.Default,
```
Comment (DanySK) [inline alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/launchers/GraphQLServerLauncher.kt:41]: add a secondary constructor taking a string and selecting the Dispatcher based on the string

**Review by DanySK — CHANGES_REQUESTED** [review]

File: alchemist-graphql/build.gradle.kts (hunk 3 above)
Comment (S-furi) [reply alchemist-graphql/build.gradle.kts:82]: I can avoid to use it at all while there is no actual test upstream.

File: alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/launchers/GraphQLServerLauncher.kt (hunk 13 above)
Comment (S-furi) [reply alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/launchers/GraphQLServerLauncher.kt:41]: This is not easy as it seems, due to detekt complaining about using coroutines dispatchers without dependency injection. I think that @AngeloFilaseta has encountered this issue too.

File: alchemist-graphql/build.gradle.kts (hunk 3 above)
Comment (DanySK) [reply alchemist-graphql/build.gradle.kts]: ok

File: alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/launchers/GraphQLServerLauncher.kt (hunk 13 above)
Comment (DanySK) [reply alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/launchers/GraphQLServerLauncher.kt:41]: What is the detekt rule?

Comment (S-furi) [reply alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/launchers/GraphQLServerLauncher.kt:41]: It is `detektJvmMain`, it fails with the message: "Dispatcher Default/IO/Unconfined is used without dependency injection. [InjectDispatcher]"

Comment (DanySK) [reply alchemist-graphql/src/jvmMain/kotlin/it/unibo/alchemist/boundary/launchers/GraphQLServerLauncher.kt:41]: Sorry I thought it was a constructor. My bad

**Comment by S-furi** [comment]
@DanySK I see one change requested but we have resolved all conversations. Am I missing something?

**Review by DanySK — APPROVED** [review]

**Pull request merged**